llvm/lib/Target/AMDGPU/GCNRegPressure.cpp

```diff
@@ -115,7 +119,19 @@ bool GCNRegPressure::less(const GCNSubtarget &ST,
     SGPRImportant = false;
   }
 
-  // compare large regs pressure
+  // In regions with spilling, we should give prefernce to the schedule with
+  // less general RP.
+  if (Occ <= MFI.getMinWavesPerEU()) {
+    unsigned GPRPressure =
+        SGPRImportant ? getSGPRNum() : getVGPRNum(ST.hasGFX90AInsts());
+    unsigned OtherGPRPressure =
+        SGPRImportant ? O.getSGPRNum() : O.getVGPRNum(ST.hasGFX90AInsts());
+
+    if (GPRPressure != OtherGPRPressure)
+      return GPRPressure < OtherGPRPressure;
+  }
```

Comment:
Hi Valery -- 

Thanks for your comments. 

This patch doesn't really intend to answer the question "which schedule to prefer" in general, but, instead, which schedule to prefer when there is spilling.

For this patch, we would like to determine which register type which is excessive / is concerning for spilling. For this purpose, I think the current determination of SGPRImportant is okay. The only default path to GCNRegPressure::less is via `mayCauseSpilling`. GCNRegPressure::less will (now) only be entered if `WavesAfter <= MFI.getMinWavesPerEU() && isRegionWithExcessRP() `. Thus, we know occupancy is concerning, and that either SGPR or VGPR pressure is excessive (I have passed the latter as a flag). The occupancy checks will pick up on which type of register is excessive -- unless there is disagreement between before / after or if they are both excessive.

Reply:
You may have excess in SGPRs but not in VGPRs, but both SGPROcc == VGPROcc == 1. In this case it would compare VGPRs (by SGPRImportant = SGPROcc < VGPROcc) but this isn't what you want. 

BTW SGPRs can be spilled into VGPRs, right? How about translating SGPR excess into a number of VGPRs used for SGPR spilling, add that number to the number of already used VGPRs and comparing only VGPRs?

Something like NumVGPRforSGPRSpills = ceil(SGPRExcess / Wavesize)? This is very conservative of course.

I'm not sure how SGPR to VGPR spilling works, cc @arsenm.

Reply:
Hi Valery,

Please see the new iteration of the work -- this has better `SGPRImportant` handling. It builds upon "Improve ExcessRP Flagging". As mentioned below, if this approach is agreed upon (at least at a high level), I'll create a separate PR for that patch.

> How about translating SGPR excess into a number of VGPRs used for SGPR spilling, add that number to the number of already used VGPRs and comparing only VGPRs?

I  think it is best to do this separately, but it does seem like an interesting feature.